diff --git a/src/core.zig b/src/core.zig
index 880684d..aa148d2 100644
--- a/src/core.zig
+++ b/src/core.zig
@@ -95,6 +95,7 @@ pub const ByteSpan = struct {
 };
 
 pub const Layout = struct {
+    left_padding: Area,
     number_line: Area,
     buffer: Area,
 };
diff --git a/src/editor.zig b/src/editor.zig
index cfc66aa..aa407b3 100644
--- a/src/editor.zig
+++ b/src/editor.zig
@@ -35,6 +35,11 @@ pub const Config = struct {
     /// Char to denote terminal lines after end of buffer
     /// See vim's :h fillchars -> eob
     end_of_buffer_char: ?u8 = null,
+    /// Imaginary width of a buffer that should be aligned (padded on the left).
+    /// `null` means "no centering".
+    /// If >`term_width`, padding is 0,
+    /// If <`term_width`, left padding is `(term_width-centering_width)/2`
+    centering_width: ?usize = null,
 };
 
 pub const Editor = struct {
diff --git a/src/terminal.zig b/src/terminal.zig
index 4aae290..2cbf920 100644
--- a/src/terminal.zig
+++ b/src/terminal.zig
@@ -411,20 +411,32 @@ pub fn terminalSize() !Dimensions {
 
 pub fn computeLayout(term_dims: Dimensions) Layout {
     const number_line_width = 5;
+    const padding_width = if (main.editor.config.centering_width) |cw|
+        if (term_dims.width > cw) @divFloor(term_dims.width - cw, 2) else 0
+    else
+        0;
+    const occupied_width = padding_width + number_line_width;
 
     return .{
-        .number_line = .{
+        .left_padding = .{
             .pos = .{},
+            .dims = .{
+                .height = term_dims.height,
+                .width = padding_width,
+            },
+        },
+        .number_line = .{
+            .pos = .{ .col = @intCast(padding_width) },
             .dims = .{
                 .height = term_dims.height,
                 .width = number_line_width,
             },
         },
         .buffer = .{
-            .pos = .{ .col = number_line_width },
+            .pos = .{ .col = @intCast(occupied_width) },
             .dims = .{
                 .height = term_dims.height,
-                .width = term_dims.width - number_line_width,
+                .width = term_dims.width - occupied_width,
             },
         },
     };
