diff --git a/src/buffer.zig b/src/buffer.zig
index 8955c3c..d8291aa 100644
--- a/src/buffer.zig
+++ b/src/buffer.zig
@@ -65,6 +65,7 @@ pub const Buffer = struct {
     uncommitted_changes: std.ArrayList(cha.Change),
     lsp_connections: std.ArrayList(*lsp.LspConnection),
     scratch: bool = false,
+    highlights: std.ArrayList(Span),
     allocator: Allocator,
 
     pub fn init(allocator: Allocator, path: ?[]const u8, content_raw: []const u8) !Buffer {
@@ -105,6 +106,7 @@ pub const Buffer = struct {
             .uncommitted_changes = std.ArrayList(cha.Change).init(allocator),
             .lsp_connections = std.ArrayList(*lsp.LspConnection).init(allocator),
             .scratch = scratch,
+            .highlights = std.ArrayList(Span).init(allocator),
             .allocator = allocator,
         };
         _ = try self.syncFs();
@@ -150,6 +152,7 @@ pub const Buffer = struct {
         self.line_positions.deinit();
         self.line_byte_positions.deinit();
         self.indents.deinit();
+        self.highlights.deinit();
 
         for (self.history.items) |*i| {
             for (i.items) |*c| c.deinit();
@@ -604,6 +607,12 @@ pub const Buffer = struct {
         }
     }
 
+    pub fn highlight(self: *Buffer) !void {
+        for (self.lsp_connections.items) |conn| {
+            try conn.highlight();
+        }
+    }
+
     pub fn renamePrompt(self: *Buffer) !void {
         const cmd = &main.editor.command_line;
         const line = self.lineContent(@intCast(self.cursor.row));
diff --git a/src/color.zig b/src/color.zig
index 1335218..20721b7 100644
--- a/src/color.zig
+++ b/src/color.zig
@@ -90,6 +90,7 @@ pub const Attr = union(enum) {
 pub const attributes = enum {
     pub const text = &[_]Attr{.{ .fg = color.white }};
     pub const selection = &[_]Attr{.{ .bg = color.gray3 }};
+    pub const highlight = &[_]Attr{.{ .bg = color.gray3 }};
     pub const keyword = &[_]Attr{.{ .fg = color.magenta }};
     pub const string = &[_]Attr{.{ .fg = color.green }};
     pub const literal = &[_]Attr{.{ .fg = color.yellow }};
diff --git a/src/lsp.zig b/src/lsp.zig
index 8ed46b8..fcd3676 100644
--- a/src/lsp.zig
+++ b/src/lsp.zig
@@ -9,6 +9,7 @@ const buf = @import("buffer.zig");
 const cha = @import("change.zig");
 const core = @import("core.zig");
 const Cursor = core.Cursor;
+const Span = core.Span;
 const fs = @import("fs.zig");
 const log = @import("log.zig");
 const main = @import("main.zig");
@@ -119,6 +120,7 @@ pub const LspConnection = struct {
                     .rename = .{
                         .prepareSupport = true,
                     },
+                    .documentHighlight = .{},
                 },
                 .workspace = .{
                     .workspaceFolders = true,
@@ -190,6 +192,8 @@ pub const LspConnection = struct {
                         try self.handleHoverResponse(arena.allocator(), response_result);
                     } else if (std.mem.eql(u8, method, "textDocument/rename")) {
                         try self.handleRenameResponse(arena.allocator(), response_result);
+                    } else if (std.mem.eql(u8, method, "textDocument/documentHighlight")) {
+                        try self.handleHighlightResponse(arena.allocator(), response_result);
                     }
                 },
                 .notification => |notif| {
@@ -236,6 +240,14 @@ pub const LspConnection = struct {
         });
     }
 
+    pub fn highlight(self: *LspConnection) !void {
+        const buffer = main.editor.active_buffer;
+        try self.sendRequest("textDocument/documentHighlight", .{
+            .textDocument = .{ .uri = buffer.uri },
+            .position = buffer.cursor.toLsp(),
+        });
+    }
+
     pub fn rename(self: *LspConnection, new_name: []const u8) !void {
         const buffer = main.editor.active_buffer;
         try self.sendRequest("textDocument/rename", .{
@@ -513,6 +525,21 @@ pub const LspConnection = struct {
         try main.editor.applyWorkspaceEdit(result.value);
     }
 
+    fn handleHighlightResponse(self: *LspConnection, arena: Allocator, resp: ?std.json.Value) !void {
+        _ = self;
+        if (resp == null or resp.? == .null) return;
+        const result = try std.json.parseFromValue([]const types.DocumentHighlight, arena, resp.?, .{});
+        const buffer = main.editor.active_buffer;
+        buffer.highlights.clearRetainingCapacity();
+        for (result.value) |hi| {
+            try buffer.highlights.append(Span.fromLsp(hi.range));
+        }
+        if (buffer.highlights.items.len > 0) {
+            log.debug(@This(), "got {} highlights\n", .{buffer.highlights.items.len});
+            main.editor.dirty.draw = true;
+        }
+    }
+
     fn handleNotification(self: *LspConnection, arena: Allocator, notif: lsp.JsonRPCMessage.Notification) !void {
         _ = self;
         // log.debug(@This(), "notification: {s}\n", .{notif.method});
diff --git a/src/main.zig b/src/main.zig
index d6b06e8..9f69106 100644
--- a/src/main.zig
+++ b/src/main.zig
@@ -410,6 +410,13 @@ fn startEditor(allocator: std.mem.Allocator) !void {
         } else if (editor.dirty.cursor) {
             editor.dirty.cursor = false;
             try term.updateCursor();
+
+            if (buffer.highlights.items.len > 0) {
+                buffer.highlights.clearRetainingCapacity();
+                // redraw next frame to clear invalid highlights
+                editor.dirty.draw = true;
+            }
+            try buffer.highlight();
         }
         perf.draw = timer.lap();
 
diff --git a/src/terminal.zig b/src/terminal.zig
index 566bb71..ca504a7 100644
--- a/src/terminal.zig
+++ b/src/terminal.zig
@@ -218,6 +218,14 @@ pub const Terminal = struct {
                     }
                 }
 
+                if (main.editor.mode == .normal) {
+                    for (buffer.highlights.items) |hi| {
+                        if (hi.inRange(.{ .row = buffer_row, .col = buffer_col })) {
+                            try co.attributes.write(co.attributes.highlight, attrs_stream.writer());
+                        }
+                    }
+                }
+
                 if (buffer.diagnostics.items.len > 0) {
                     for (buffer.diagnostics.items) |diagnostic| {
                         const span = diagnostic.span;
