diff --git c/src/buffer.zig w/src/buffer.zig
index f8ac475..b3d4056 100644
--- c/src/buffer.zig
+++ w/src/buffer.zig
@@ -171,14 +171,6 @@ pub const Buffer = struct {
         _ = try self.syncFs();
 
         self.file_history_index = self.history_index;
-
-        const msg = try std.fmt.allocPrint(
-            self.allocator,
-            "{s} {}B written",
-            .{ self.path, self.content_raw.items.len },
-        );
-        defer self.allocator.free(msg);
-        try main.editor.sendMessage(msg);
     }
 
     pub fn moveCursor(self: *Buffer, new_cursor: Cursor) !void {
@@ -361,6 +353,11 @@ pub const Buffer = struct {
         self.history_index = self.history.items.len - 1;
 
         try main.editor.dotRepeatCommitReady();
+
+        if (main.editor.config.autosave) {
+            try self.write();
+            log.debug(@This(), "autosave {s}\n", .{self.path});
+        }
     }
 
     pub fn changeInsertText(self: *Buffer, text: []const u21) !void {
diff --git c/src/editor.zig w/src/editor.zig
index 21fc180..6e697b6 100644
--- c/src/editor.zig
+++ w/src/editor.zig
@@ -27,6 +27,7 @@ pub const DotRepeat = enum {
 };
 
 pub const Config = struct {
+    autosave: bool = false,
 };
 
 pub const Editor = struct {
diff --git c/src/main.zig w/src/main.zig
index 285f63d..d590f7b 100644
--- c/src/main.zig
+++ w/src/main.zig
@@ -331,6 +331,13 @@ fn startEditor(allocator: std.mem.Allocator) !void {
 
                     if (editor.mode == .normal and eql(u8, multi_key, " w")) {
                         try buffer.write();
+                        const msg = try std.fmt.allocPrint(
+                            allocator,
+                            "{s} {}B written",
+                            .{ buffer.path, buffer.content_raw.items.len },
+                        );
+                        defer allocator.free(msg);
+                        try editor.sendMessage(msg);
                     } else if (editor.mode == .normal and eql(u8, multi_key, " d")) {
                         try buffer.goToDefinition();
                     } else if (editor.mode == .normal and eql(u8, multi_key, " r")) {
