diff --git a/src/buffer.zig b/src/buffer.zig
index ea5bb1c..3a50471 100644
--- a/src/buffer.zig
+++ b/src/buffer.zig
@@ -167,14 +167,6 @@ pub const Buffer = struct {
         _ = try self.syncFs();
 
         self.file_history_index = self.history_index;
-
-        const msg = try std.fmt.allocPrint(
-            self.allocator,
-            "{s} {}B written",
-            .{ self.path, self.content_raw.items.len },
-        );
-        defer self.allocator.free(msg);
-        try main.editor.sendMessage(msg);
     }
 
     pub fn moveCursor(self: *Buffer, new_cursor: Cursor) void {
@@ -363,6 +355,11 @@ pub const Buffer = struct {
         self.history_index = self.history.items.len - 1;
 
         main.editor.dotRepeatCommitReady();
+
+        if (main.editor.config.autosave) {
+            log.debug(@This(), "autosave {s}\n", .{self.path});
+            self.write() catch |e| log.err(@This(), "write buffer error: {}", .{e});
+        }
     }
 
     pub fn changeInsertText(self: *Buffer, text: []const u21) FatalError!void {
@@ -494,7 +491,7 @@ pub const Buffer = struct {
         }
     }
 
-    pub fn undo(self: *Buffer) !void {
+    pub fn undo(self: *Buffer) FatalError!void {
         log.debug(@This(), "undo: {?}/{}\n", .{ self.history_index, self.history.items.len });
         if (self.history_index) |h_idx| {
             const hist_to_undo = self.history.items[h_idx].items;
@@ -506,6 +503,11 @@ pub const Buffer = struct {
                 self.moveCursor(inv_change.new_span.?.start);
             }
             self.history_index = if (h_idx > 0) h_idx - 1 else null;
+
+            if (main.editor.config.autosave) {
+                self.write() catch |e| log.err(@This(), "write buffer error: {}", .{e});
+                log.debug(@This(), "autosave {s}\n", .{self.path});
+            }
         }
     }
 
@@ -521,6 +523,11 @@ pub const Buffer = struct {
             self.moveCursor(change.new_span.?.start);
         }
         self.history_index = redo_idx;
+
+        if (main.editor.config.autosave) {
+            self.write() catch |e| log.err(@This(), "write buffer error: {}", .{e});
+            log.debug(@This(), "autosave {s}\n", .{self.path});
+        }
     }
 
     pub fn textAt(self: *const Buffer, span: Span) []const u21 {
diff --git a/src/editor.zig b/src/editor.zig
index 33b09f3..3747f4f 100644
--- a/src/editor.zig
+++ b/src/editor.zig
@@ -32,6 +32,7 @@ pub const DotRepeat = enum {
 };
 
 pub const Config = struct {
+    autosave: bool = false,
     /// Char to denote terminal lines after end of buffer
     /// See vim's :h fillchars -> eob
     end_of_buffer_char: ?u8 = null,
